zerver/lib/default_streams.py

```diff
@@ -0,0 +1,35 @@
+from typing import List, Set
+
+from zerver.lib.types import APIStreamDict
+from zerver.models import DefaultStream, Stream
+
+
+def get_slim_realm_default_streams(realm_id: int) -> List[Stream]:
+    # This slightly convoluted helper tells the Django ORM to get all the
+    # default streams in a single query without prematurely joining to other
+    # tables, since all of our callers just need values from Stream itself.  It
+    # also prevents subsequent round trips to the database that can happen with
+    # more naive constructions (even to get Stream objects).
+    #
+    # The above paradigm is enforced by at least three tests that verify query
+    # counts, and test_query_count in test_subs.py makes sure that the query
+    # itself is not like 11000 bytes, which is what you would get if you did
+    # select_related().  Please avoid select_related() with streams, because
+    # you will get all kinds of query bloat related to Realm and UserGroup.
+    stream_ids = DefaultStream.objects.filter(realm_id=realm_id).values_list("stream_id")
+    streams = Stream.objects.filter(id__in=stream_ids)
```

Comment:
Although [`select_related()`](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#select-related) joins everything by default, you can use it selectively:

```python
DefaultStream.objects.select_related("stream").filter(realm_id=realm_id)
```

This results in a sane query like

`SELECT "zerver_defaultstream"."id", "zerver_defaultstream"."realm_id", "zerver_defaultstream"."stream_id", "zerver_stream"."id", "zerver_stream"."name", "zerver_stream"."realm_id", "zerver_stream"."date_created", "zerver_stream"."deactivated", "zerver_stream"."description", "zerver_stream"."rendered_description", "zerver_stream"."recipient_id", "zerver_stream"."invite_only", "zerver_stream"."history_public_to_subscribers", "zerver_stream"."is_web_public", "zerver_stream"."stream_post_policy", "zerver_stream"."is_in_zephyr_realm", "zerver_stream"."email_token", "zerver_stream"."message_retention_days", "zerver_stream"."can_remove_subscribers_group_id", "zerver_stream"."first_message_id" FROM "zerver_defaultstream" INNER JOIN "zerver_stream" ON ("zerver_defaultstream"."stream_id" = "zerver_stream"."id") WHERE "zerver_defaultstream"."realm_id" = 2`

and saves a database round trip.

Reply:
~~~
 19     streams = Stream.objects.filter(
 20         id__in=DefaultStream.objects.filter(realm_id=realm_id).values_list("stream_id")
 21     )